File: web/src/levels.js
// Experience and levels — the same arithmetic as the Android app's Levels.kt, integer
// for integer. A float here would eventually disagree with the phone on a boundary, and
// the two versions are meant to describe the same profile.

export const MAX_LEVEL = 100;
export const XP_WIN = 25;
export const XP_LOSS = 10;

/** What the first level costs. */
const BASE = 20;

/** How much more each level costs than the one before it. */
const STEP = 5;

/** Experience needed to go from `level` to the next one. Zero at the top. */
export function costOf(level) {
  return (level < 1 || level >= MAX_LEVEL) ? 0 : BASE + STEP * (level - 1);
}

/** Total experience needed to reach `level` from scratch. Closed form, not a loop. */
export function totalTo(level) {
  const done = Math.min(Math.max(level, 1), MAX_LEVEL) - 1;
  return done * BASE + STEP * done * (done - 1) / 2;
}

export const xpFor = (won) => (won ? XP_WIN : XP_LOSS);

/** The level this much experience buys, clamped to the top. */
export function levelAt(xp) {
  if (xp <= 0) return 1;
  let level = 1;
  let spent = 0;
  while (level < MAX_LEVEL) {
    const cost = costOf(level);
    if (xp - spent < cost) break;
    spent += cost;
    level++;
  }
  return level;
}

/** How far into the current level, in points. */
export const into = (xp) => Math.max(xp, 0) - totalTo(levelAt(xp));

/** How wide the current level is, in points. Zero at the top. */
export const span = (xp) => costOf(levelAt(xp));

/** Progress through the current level, 0..100. A maxed profile reads 100. */
export function percent(xp) {
  const width = span(xp);
  return width <= 0 ? 100 : Math.trunc((into(xp) * 100) / width);
}

/** Experience still owed before the next level. Zero at the top. */
export function toNext(xp) {
  const width = span(xp);
  return width <= 0 ? 0 : width - into(xp);
}

/** Everything from scratch to the top. */
export const fullRun = () => totalTo(MAX_LEVEL);

File: app/src/main/java/com/zknw/unoduo/progress/Levels.kt
package com.zknw.unoduo.progress

/**
 * Experience and levels.
 *
 * A finished round is worth a flat amount — winning more than losing, but losing is
 * never worth nothing, because a level that only moves when you win punishes exactly
 * the players who need the encouragement.
 *
 * The curve is linear on purpose: each level costs a fixed amount more than the one
 * before it. An exponential curve makes the first ten levels meaningless and the last
 * ten unreachable; a linear one keeps every level a step you can actually see coming.
 *
 * Everything here is integer arithmetic on purpose. The same numbers are computed on
 * the phone and in the browser, and a float would eventually disagree on a boundary.
 */
object Levels {

    /** The last level. Experience past it is kept but buys nothing. */
    const val MAX = 100

    const val XP_WIN = 25
    const val XP_LOSS = 10

    /** What the first level costs. */
    private const val BASE = 20

    /** How much more each level costs than the one before it. */
    private const val STEP = 5

    /** Experience needed to go from [level] to the next one. Zero at [MAX]. */
    fun costOf(level: Int): Int =
        if (level < 1 || level >= MAX) 0 else BASE + STEP * (level - 1)

    /**
     * Total experience needed to reach [level] from scratch.
     *
     * Closed form rather than a loop: this is called on every frame that draws the
     * experience bar.
     */
    fun totalTo(level: Int): Int {
        val done = level.coerceIn(1, MAX) - 1
        return done * BASE + STEP * done * (done - 1) / 2
    }

    /** Experience for a finished round. */
    fun xpFor(won: Boolean): Int = if (won) XP_WIN else XP_LOSS

    /** The level [xp] buys, clamped to [MAX]. */
    fun levelAt(xp: Int): Int {
        if (xp <= 0) return 1
        var level = 1
        var spent = 0
        while (level < MAX) {
            val cost = costOf(level)
            if (xp - spent < cost) break
            spent += cost
            level++
        }
        return level
    }

    /** How far into the current level [xp] is, in points. */
    fun into(xp: Int): Int = xp.coerceAtLeast(0) - totalTo(levelAt(xp))

    /** How wide the current level is, in points. Zero at [MAX]. */
    fun span(xp: Int): Int = costOf(levelAt(xp))

    /** Progress through the current level, 0..100. A maxed profile reads 100. */
    fun percent(xp: Int): Int {
        val width = span(xp)
        return if (width <= 0) 100 else (into(xp) * 100) / width
    }

    /** Experience still owed before the next level. Zero at [MAX]. */
    fun toNext(xp: Int): Int {
        val width = span(xp)
        return if (width <= 0) 0 else width - into(xp)
    }

    /** Everything from scratch to [MAX], for the profile to show what it is in for. */
    val fullRun: Int get() = totalTo(MAX)
}
